### DISCUSSION TITLE
VITS training - why eval loss converges quickly and continues to increase?

### DISCUSSION THREAD

**Discussion by nicemanis** [coqui-ai/TTS#3377] [General Q&A]
I've been experimenting with VITS fine-tuning on my own datasets from a VITS VCTK checkpoint. While I managed to get somewhat decent results, it still had problems with the speech flow and pauses between words, similar to the issues reported in this discussion https://github.com/coqui-ai/TTS/discussions/2834. For all of my training runs I observed that the eval loss initially rapidly decreased, but then continued to increase after 10-20K steps, suggesting overfitting.

In order to isolate the problem and verify the validity of the training recipe, I tried to train VITS from scratch on the VCTK dataset; however it seems that I'm encountering the same problem.

## Tensorboard logs
![image](https://github.com/coqui-ai/TTS/assets/9463816/098595c5-97fe-4a47-8bdc-078607a8491f)
![image](https://github.com/coqui-ai/TTS/assets/9463816/ba341ba7-3dac-4daa-a825-651a021a32c9)
![image](https://github.com/coqui-ai/TTS/assets/9463816/89d5dd6b-cd33-4154-9881-8afa7467e42d)

## config.json
```json
{
    "output_path": "",
    "logger_uri": null,
    "run_name": "",
    "project_name": null,
    "run_description": "",
    "print_step": 50,
    "plot_step": 100,
    "model_param_stats": false,
    "wandb_entity": null,
    "dashboard_logger": "tensorboard",
    "log_model_step": 1000,
    "save_step": 5000,
    "save_n_checkpoints": 5,
    "save_checkpoints": true,
    "save_all_best": false,
    "save_best_after": 10000,
    "target_loss": null,
    "print_eval": false,
    "test_delay_epochs": -1,
    "run_eval": true,
    "run_eval_steps": null,
    "distributed_backend": "nccl",
    "distributed_url": "tcp://localhost:54321",
    "mixed_precision": false,
    "epochs": 1000,
    "batch_size": 32,
    "eval_batch_size": 32,
    "grad_clip": [
        1000.0,
        1000.0
    ],
    "scheduler_after_epoch": true,
    "lr": 0.001,
    "optimizer": "AdamW",
    "optimizer_params": {
        "betas": [
            0.8,
            0.99
        ],
        "eps": 1e-09,
        "weight_decay": 0.01
    },
    "lr_scheduler": null,
    "lr_scheduler_params": {},
    "use_grad_scaler": false,
    "cudnn_enable": true,
    "cudnn_deterministic": false,
    "cudnn_benchmark": false,
    "training_seed": 54321,
    "model": "vits",
    "num_loader_workers": 8,
    "num_eval_loader_workers": 8,
    "use_noise_augment": false,
    "audio": {
        "fft_size": 1024,
        "sample_rate": 22050,
        "win_length": 1024,
        "hop_length": 256,
        "num_mels": 80,
        "mel_fmin": 0,
        "mel_fmax": null
    },
    "use_phonemes": true,
    "phonemizer": "multi_phonemizer",
    "phoneme_language": null,
    "compute_input_seq_cache": true,
    "text_cleaner": "multilingual_cleaners",
    "enable_eos_bos_chars": false,
    "test_sentences_file": "",
    "phoneme_cache_path": "/tmp/coqui-tts/phoneme_cache",
    "characters": {
        "characters_class": "TTS.tts.utils.text.characters.IPAPhonemes",
        "vocab_dict": null,
        "pad": "<PAD>",
        "eos": "<EOS>",
        "bos": "<BOS>",
        "blank": "<BLNK>",
        "characters": "iy\u0268\u0289\u026fu\u026a\u028f\u028ae\u00f8\u0258\u0259\u0275\u0264o\u025b\u0153\u025c\u025e\u028c\u0254\u00e6\u0250a\u0276\u0251\u0252\u1d7b\u0298\u0253\u01c0\u0257\u01c3\u0284\u01c2\u0260\u01c1\u029bpbtd\u0288\u0256c\u025fk\u0261q\u0262\u0294\u0274\u014b\u0272\u0273n\u0271m\u0299r\u0280\u2c71\u027e\u027d\u0278\u03b2fv\u03b8\u00f0sz\u0283\u0292\u0282\u0290\u00e7\u029dx\u0263\u03c7\u0281\u0127\u0295h\u0266\u026c\u026e\u028b\u0279\u027bj\u0270l\u026d\u028e\u029f\u02c8\u02cc\u02d0\u02d1\u028dw\u0265\u029c\u02a2\u02a1\u0255\u0291\u027a\u0267\u02b2\u025a\u02de\u026b",
        "punctuations": "!'(),-.:;? ",
        "phonemes": null,
        "is_unique": false,
        "is_sorted": true
    },
    "add_blank": true,
    "batch_group_size": 0,
    "loss_masking": null,
    "min_audio_len": 22050,
    "max_audio_len": 220500,
    "min_text_len": 1,
    "max_text_len": Infinity,
    "compute_f0": false,
    "compute_energy": false,
    "compute_linear_spec": true,
    "precompute_num_workers": 0,
    "start_by_longest": false,
    "shuffle": false,
    "drop_last": false,
    "datasets": [
        {
            "formatter": "vctk",
            "dataset_name": "vctk",
            "path": "/tmp/coqui-tts/VCTK",
            "meta_file_train": "",
            "ignored_speakers": [
                "p261",
                "p225",
                "p294",
                "p347",
                "p238",
                "p234",
                "p248",
                "p335",
                "p245",
                "p326",
                "p302"
            ],
            "language": "en",
            "phonemizer": "",
            "meta_file_val": "",
            "meta_file_attn_mask": ""
        }
    ],
    "test_sentences": [
        [
            "It took me quite a long time to develop a voice, and now that I have it I'm not going to be silent.",
            "VCTK_p277",
            null,
            "en"
        ]
    ],
    "eval_split_max_size": null,
    "eval_split_size": 0.01,
    "use_speaker_weighted_sampler": false,
    "speaker_weighted_sampler_alpha": 1.0,
    "use_language_weighted_sampler": true,
    "language_weighted_sampler_alpha": 1.0,
    "use_length_weighted_sampler": false,
    "length_weighted_sampler_alpha": 1.0,
    "model_args": {
        "num_chars": 131,
        "out_channels": 513,
        "spec_segment_size": 32,
        "hidden_channels": 192,
        "hidden_channels_ffn_text_encoder": 768,
        "num_heads_text_encoder": 2,
        "num_layers_text_encoder": 6,
        "kernel_size_text_encoder": 3,
        "dropout_p_text_encoder": 0.1,
        "dropout_p_duration_predictor": 0.5,
        "kernel_size_posterior_encoder": 5,
        "dilation_rate_posterior_encoder": 1,
        "num_layers_posterior_encoder": 16,
        "kernel_size_flow": 5,
        "dilation_rate_flow": 1,
        "num_layers_flow": 4,
        "resblock_type_decoder": "1",
        "resblock_kernel_sizes_decoder": [
            3,
            7,
            11
        ],
        "resblock_dilation_sizes_decoder": [
            [
                1,
                3,
                5
            ],
            [
                1,
                3,
                5
            ],
            [
                1,
                3,
                5
            ]
        ],
        "upsample_rates_decoder": [
            8,
            8,
            2,
            2
        ],
        "upsample_initial_channel_decoder": 512,
        "upsample_kernel_sizes_decoder": [
            16,
            16,
            4,
            4
        ],
        "periods_multi_period_discriminator": [
            2,
            3,
            5,
            7,
            11
        ],
        "use_sdp": true,
        "noise_scale": 1.0,
        "inference_noise_scale": 0.667,
        "length_scale": 1.0,
        "noise_scale_dp": 1.0,
        "inference_noise_scale_dp": 1.0,
        "max_inference_len": null,
        "init_discriminator": true,
        "use_spectral_norm_disriminator": false,
        "use_speaker_embedding": true,
        "num_speakers": 98,
        "speakers_file": "",
        "d_vector_file": null,
        "speaker_embedding_channels": 256,
        "use_d_vector_file": false,
        "d_vector_dim": 0,
        "detach_dp_input": true,
        "use_language_embedding": true,
        "embedded_language_dim": 4,
        "num_languages": 1,
        "language_ids_file": "",
        "use_speaker_encoder_as_loss": false,
        "speaker_encoder_config_path": "",
        "speaker_encoder_model_path": "",
        "condition_dp_on_speaker": true,
        "freeze_encoder": false,
        "freeze_DP": false,
        "freeze_PE": false,
        "freeze_flow_decoder": false,
        "freeze_waveform_decoder": false,
        "encoder_sample_rate": null,
        "interpolate_z": true,
        "reinit_DP": false,
        "reinit_text_encoder": false
    },
    "lr_gen": 0.0002,
    "lr_disc": 0.0002,
    "lr_scheduler_gen": "ExponentialLR",
    "lr_scheduler_gen_params": {
        "gamma": 0.999875,
        "last_epoch": -1
    },
    "lr_scheduler_disc": "ExponentialLR",
    "lr_scheduler_disc_params": {
        "gamma": 0.999875,
        "last_epoch": -1
    },
    "kl_loss_alpha": 1.0,
    "disc_loss_alpha": 1.0,
    "gen_loss_alpha": 1.0,
    "feat_loss_alpha": 1.0,
    "mel_loss_alpha": 45.0,
    "dur_loss_alpha": 1.0,
    "speaker_encoder_loss_alpha": 1.0,
    "return_wav": true,
    "use_weighted_sampler": false,
    "weighted_sampler_attrs": {},
    "weighted_sampler_multipliers": {},
    "r": 1,
    "num_speakers": 98,
    "use_speaker_embedding": true,
    "speakers_file": "",
    "speaker_embedding_channels": 256,
    "language_ids_file": "",
    "use_language_embedding": true,
    "use_d_vector_file": false,
    "d_vector_file": null,
    "d_vector_dim": 0
}
```

I compared my config with the one from the pre-trained VITS VCTK. The only notable differences that I noticed were:
- My config has `grad_clip: [1000.0, 1000.0]` instead of `grad_clip: [5.0, 5.0]`
- The pre-trained model's config has a bunch of additional paramaters for the audio config, I'm not sure if those are default values or something that has been adjusted
- I use IPAPhonemes instead of VitsCharacters, but both configs use Espeak phonemizer
- The pre-trained model has `max_audio_len` set to Infinity
- My config has `compute_energy: false`, `shuffle: false`, and `drop_last: false`
- My config has defined `periods_multi_period_discriminator`
- My config has `init_discriminator: true`
- My config has `use_speaker_embedding: true` and `use_language_embedding: true`

Based on the logs, there seems to be a problem with the feats and KL loss.
Any ideas how to fix this issue? Should I adjust the learning rate, although the pre-trained had the same settings? Maybe there is some othere issue with the config? Currently, the training is at around 185K steps and the best model is saved at 80K steps.

**Comment by erogol** [comment]
Check the audio quality. Eval loss is not always the best metric for the final quality.
